### PR TITLE
png: add support for 16-bit PNG files.

### PR DESCRIPTION
**Pull request by krasin**
Note: libpng.load now returns two arguments. The first argument is the result tensor,
and the second argument is the color bitness. It will be either 8 or 16. For images
with smaller color bit depth (for example, 4), the value is upgraded to 8 bits.

See also #74 and #75.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -118,8 +118,9 @@ local function decompress(tensor, depth, tensortype)
 end
 rawset(image, 'decompress', decompress)
 
-local function processPNG(img, depth, tensortype)
+local function processPNG(img, depth, bit_depth, tensortype)
```
Comment (soumith) [inline init.lua]: make bit_depth to be the last argument, so as to not break existing scripts

Comment (soumith) [inline init.lua]: oh nevermind, no one uses this directly.

**Comment by soumith** [comment]
this looks pretty good to me overall

**Comment by soumith** [comment]
have to tested this on both 8-bit and 16-bit stuff? can you add some simple tests (even just to display a couple of images for sanity checks) to the test folder

**Comment by krasin** [comment]
@soumith thanks for your comments.

re: tests. I would like to add PNG tests to [torch/image/test](https://github.com/torch/image/tree/master/test) dir, just like you have for PPM, PGM and JPG.

One thing that I failed to figure out myself, is how to run these tests. I don't see them hooked up in [CMakeLists.txt](https://github.com/torch/image/blob/master/CMakeLists.txt). Am I missing something?

**Comment by soumith** [comment]
we just run them manually with the interpreter. For example:
th test_compress_jpg.lua

**Comment by krasin** [comment]
@soumith I have made the first draft for test_png.lua, please take a look at the commit above.

I generate real PNG files of a very small size and load them with different options:

```
gray16-1x2.png: PNG image data, 1 x 2, 16-bit grayscale, non-interlaced
gray3x1.png:    PNG image data, 3 x 1, 8-bit grayscale, non-interlaced
```

I also plan to  cover color images, but I would like to hear your feedback on the test code first. Is there a better way to specify a Tensor literal?

File: test/test_png.lua [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+function checkPNG(imfile, depth, tensortype, want)
+  local img = image.load(imfile, 1, tensortype)
+  -- Tensors have to be converted to double, since assertTensorEq does not support ByteTensor
+  mytester:assertTensorEq(img:double(), want:double(), precision_mean, string.format('%s: pixel values are unexpected', imfile))
+end
+
+function test.LoadPNG()
+  -- Gray 8-bit PNG image with width = 3, height = 1
+  local gray8byte = torch.ByteTensor(1, 1, 3)
```
Comment (soumith) [inline test/test_png.lua]: torch.ByteTensor({{{0,127,255}}})

File: test/test_png.lua [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  gray8byte[1][1][1] = 0
+  gray8byte[1][1][2] = 127
+  gray8byte[1][1][3] = 255
+  checkPNG('gray3x1.png', 1, 'byte', gray8byte)
+
+  local gray8double = torch.DoubleTensor(1, 1, 3)
```
Comment (soumith) [inline test/test_png.lua]: torch.ByteTensor({{{0,127/255,1}}})

File: test/test_png.lua [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  gray8double[1][1][1] = 0
+  gray8double[1][1][2] = 127 / 255
+  gray8double[1][1][3] = 1
+  checkPNG('gray3x1.png', 1, 'double', gray8double)
+
+
+  -- Gray 16-bit PNG image with width=1, height = 2
+  local gray16byte = torch.ByteTensor(1, 2, 1)
```
Comment (soumith) [inline test/test_png.lua]: torch.ByteTensor({{{0},{255}}})

**Comment by soumith** [comment]
yes, gave example in in-line comments.

File: test/test_png.lua (hunk 2 above)
Comment (krasin) [inline test/test_png.lua]: Thanks, done.

File: test/test_png.lua (hunk 3 above)
Comment (krasin) [inline test/test_png.lua]: Done.

File: test/test_png.lua (hunk 4 above)
Comment (krasin) [inline test/test_png.lua]: Done.

**Comment by krasin** [comment]
@soumith thank you!

I have added all the tests I wanted (8-bit / 16-bit, gray / RGB) and fixed a bug with 16-bit RGB images in the process of doing that.

Please, take another look!

**Comment by krasin** [comment]
Hi @soumith,

is there anything I can do to improve this CL?
I would appreciate any feedback.

Thanks! :)

**Comment by soumith** [comment]
no, this looks perfect, I'm sorry for the delayed response, sometimes things get lost in the haystack. thanks again for the PR.

**Comment by krasin** [comment]
@soumith thanks!

**Pull request merged**